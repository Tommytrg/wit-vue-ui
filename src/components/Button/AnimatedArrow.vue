<template>
  <div ref="arrow">
    <WIcon :name="IconName.RightArrow" class="arrow ml-md h-[14px]" :color="color"></WIcon>
  </div>
</template>

<script setup lang="ts">
import { gsap } from 'gsap'
import { computed, ref, watch } from 'vue'

import WIcon from '../icon/WIcon.vue'
import { IconName } from '../icon/WIcon'

const arrow = ref(null)

const props = defineProps({
  hover: {
    type: Boolean,
    default: false
  },
  color: {
    type: String,
    default: '#fff'
  }
})

const hover = computed(() => props.hover)

watch(hover, (value) => {
  if (value) {
    gsap.to(arrow.value, {
      translateX: 8,
      ease: 'circ.out',
      yoyo: true,
      repeat: -1
    })
  } else {
    gsap.killTweensOf(arrow.value)
    gsap.to(arrow.value, {
      translateX: 0
    })
  }
})
</script>
