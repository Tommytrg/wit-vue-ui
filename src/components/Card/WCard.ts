export enum CardType {
  base = 'base',
  icon = 'icon',
  link = 'link',
}

export const cardTypes = Object.values(CardType)