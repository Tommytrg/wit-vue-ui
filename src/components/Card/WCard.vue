<template>
  <div v-if="isLinkCard">
    <a
      :href="url"
      :aria-label="title"
      target="_blank"
      class="cursor-pointer"
      @mouseenter="toggleHover"
      @mouseleave="toggleHover"
    >
      <div
        class="card border rounded-lg px-xl py-xl mr-sm mb-sm max-w-md h-full"
      >
        <div class="flex justify-between items-center">
          <h3 class="text-xl text-black-950 font-semibold leading-4">
            {{ title }}
          </h3>
          <AnimatedArrow :hover="hover" color="#C6C6C6" />
        </div>
        <hr class="hr my-md" />
        <slot name="description" class="text-sm text-black-950"></slot>
      </div>
    </a>
  </div>
  <div
    v-else class="h-full w-auto out-of-boundaries card border-2 border-black-950 bg-white-50 rounded-lg px-xl py-xl max-w-sm"
  >
  <div v-if="isBaseCard">
    <slot name="content"></slot>
  </div>
  <div v-if="isIconCard">
    <RoundedIcon class="rounded-icon" icon-size="60px">
      <slot name="icon">Icon</slot>
    </RoundedIcon>
    <div class="ml-xl px-sm">
      <h3 class="text-2xl text-black-950 font-semibold leading-4 pl-14">
        {{ title }}
      </h3>
      <hr class="hr my-md ml-14" />
    </div>
    <p class="text text-black-950 mb-lg">
      {{ description }}
    </p>

    <SlashesIcon class="slashes w-2xl sm:hidden"></SlashesIcon>
    <a :href="url" :aria-label="urlLabel" target="_blank">
      <WButton class="action" type="arrow">{{ urlLabel }}</WButton>
    </a>
  </div>
  </div>
</template>

<script setup lang="ts">
import { cardTypes, CardType } from './WCard.ts'
import { PropType, computed, ref } from 'vue'
import AnimatedArrow from '../AnimatedArrow.vue'
import SlashesIcon from '@/assets/svg/slashes.svg?component'
import WButton from '../Button/WButton.vue'
import RoundedIcon from '../RoundedIcon.vue'
const hover = ref(false)
const props = defineProps({
  type: {
    type: String as PropType<CardType>,
    default: CardType.Base,
    validator(value: CardType) {
      return cardTypes.includes(value)
    }
  },
  title: {
    type: String,
    default: 'Title',
  },
  description: {
    type: String,
    default: 'Description',
  },
  urlLabel: {
    type: String,
    required: false,
  },
  url: {
    type: String,
    required: false,
  },
})
const isIconCard = computed(() => props.type === CardType.Icon)
const isLinkCard = computed(() => props.type === CardType.Link)
const isBaseCard = computed(() => props.type === CardType.Base)
function toggleHover() {
  hover.value = !hover.value
}
</script>

<style scoped lang="scss">
:slotted(svg) {
  width: 60px;
  height: 60px;
}

.out-of-boundaries {
  // add margin equal to out of boundaries
  margin: 15px 30px 25px 15px;
}

.rounded-icon {
  position: absolute;
  top: -15px;
  left: -15px;
}

.action {
  position: absolute;
  right: -30px;
  bottom: 16px;
}

.slashes {
  position: absolute;
  bottom: 32px;
}

.card {
  box-shadow: 7px 10px 0px rgb(0, 0, 0);
  position: relative;
}

.title-h2 {
  font-weight: bold;
}

.hr {
  height: 3px; /* Modern Browsers */
  @apply bg-black-950;
}

.circle {
  line-height: 1;
  font-size: 24px;
  height: 48px;
  min-width: 48px;
  align-content: center;
  text-align: center;
  @apply bg-wit-blue-500;
}
</style>

