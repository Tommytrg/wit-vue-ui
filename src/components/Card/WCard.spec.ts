import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import WCard from './WCard.vue'
import { CardType } from './WCard'
import AnimatedArrow from '../AnimatedArrow.vue'

describe('WCard', () => {
  it('base', () => {
    const wrapper = mount(WCard, {
      props: {
        type: CardType.base
      },
      slots: {
        default: 'Button'
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('icon', () => {
    const wrapper = mount(WCard, {
      props: {
        type: CardType.icon,
        title: 'Custom title',
        description: 'Custom description',
        url: '#url',
        urlLabel: 'Link to'
      },
      slots: {
        icon: 'WIcon'
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('link', () => {
    const wrapper = mount(WCard, {
      props: {
        type: CardType.link,
        title: 'Custom link title',
      },
      slots: {
        description: '<p>Link description</p>'
      }
    })

    const found = wrapper.findComponent(AnimatedArrow)
    expect(found.exists()).toBe(true)
  })
})
