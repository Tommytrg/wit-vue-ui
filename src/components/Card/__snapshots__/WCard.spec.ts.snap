// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WCard > base 1`] = `
<div
  class="h-full w-auto out-of-boundaries card border-2 border-black-950 bg-white-50 rounded-lg px-xl py-xl max-w-sm"
  data-v-47e965dc=""
>
  <div
    data-v-47e965dc=""
  >
    
    
  </div>
  <!--v-if-->
</div>
`;

exports[`WCard > icon 1`] = `
<div
  class="h-full w-auto out-of-boundaries card border-2 border-black-950 bg-white-50 rounded-lg px-xl py-xl max-w-sm"
  data-v-47e965dc=""
>
  <!--v-if-->
  <div
    data-v-47e965dc=""
  >
    <div
      class="border border-black-950 w-min rounded-full p-md bg-wit-blue-500 rounded-icon"
      data-v-47e965dc=""
    >
      <div
        class="icon-container flex justify-center items-center"
      >
        
        
        WIcon
        
        
      </div>
    </div>
    <div
      class="ml-xl px-sm"
      data-v-47e965dc=""
    >
      <h3
        class="text-2xl text-black-950 font-semibold leading-4 pl-14"
        data-v-47e965dc=""
      >
        Custom title
      </h3>
      <hr
        class="hr my-md ml-14"
        data-v-47e965dc=""
      />
    </div>
    <p
      class="text text-black-950 mb-lg"
      data-v-47e965dc=""
    >
      Custom description
    </p>
    <svg
      class="slashes w-[68px] sm:hidden"
      data-v-47e965dc=""
      fill="none"
      height="22"
      viewBox="0 0 119 22"
      width="119"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M18.976.684.4 19.259l2.192 2.192L21.168 2.876zM35.164.686 16.588 19.26l2.192 2.192L37.357 2.878zM51.362.687 32.787 19.263l2.192 2.192L53.554 2.88zM67.557.682 48.981 19.258l2.192 2.192L69.75 2.874zM83.746.684 65.17 19.26l2.192 2.192L85.938 2.876zM99.934.686 81.358 19.262l2.192 2.192 18.576-18.576zM116.131.688 97.556 19.264l2.192 2.192L118.323 2.88z"
        fill="#00E2ED"
      />
    </svg>
    <a
      aria-label="Link to"
      data-v-47e965dc=""
      href="#url"
      target="_blank"
    >
      <button
        class="btn w-fit flex items-center btn-secondary action"
        data-v-47e965dc=""
        data-v-870c3481=""
      >
        
        Link to
        
        <div
          data-v-870c3481=""
        >
          <svg
            class="arrow ml-md h-[14px]"
            fill="none"
            height="25"
            viewBox="0 0 15 25"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.75 24.23c-.44 0-.88-.16-1.22-.49-.7-.68-.73-1.79-.05-2.5l8.66-9.04-8.66-9.04A1.769 1.769 0 0 1 4.03.71l9.83 10.26c.66.68.66 1.76 0 2.45L4.03 23.68c-.35.37-.82.55-1.28.55"
              fill="#C6C6C6"
            />
          </svg>
        </div>
      </button>
    </a>
  </div>
</div>
`;
